src/grip/Grip.tsx

```diff
@@ -0,0 +1,62 @@
+import React, { useEffect, useState } from 'react'
+
+import { A11yProps, pickA11yProps } from '../_utils/interfaces'
+import { GripHandle } from './GripHandle'
+
+export interface GripProps extends A11yProps {
+  children?: React.ReactNode
+  onSlideUp: () => void
+  onSlideDown: () => void
+  className?: string
+  disabled?: boolean
+}
+
+export const SLIDE_OFFSET = 20 // To get more precise with feeling/testing
+
+export const touchEndListener = (
+  clientY: number,
+  fingerYPosition: number,
+  setFingerYPosition: React.Dispatch<number>,
+  props: GripProps,
+) => {
+  const { onSlideDown, onSlideUp, disabled = false } = props
+  if (fingerYPosition !== null && !disabled) {
+    if (clientY < fingerYPosition - SLIDE_OFFSET) {
+      onSlideUp()
+    } else if (clientY > fingerYPosition + SLIDE_OFFSET) {
+      onSlideDown()
+    }
+    setFingerYPosition(null)
+  }
+}
+
+export const Grip = (props: GripProps): JSX.Element => {
+  const { children = null, className = '' } = props
+  const a11yAttrs = pickA11yProps(props)
+  const [fingerYPosition, setFingerYPosition] = useState<number>(null)
+
+  // Listening to finger being lift from the screen and check position
+  // difference with starting point
+  useEffect(() => {
+    const delegatedTouchEndListener = (e: TouchEvent) =>
+      touchEndListener(e.changedTouches.item(0).clientY, fingerYPosition, setFingerYPosition, props)
+    window.addEventListener('touchend', delegatedTouchEndListener)
+
+    return function cleanup() {
+      window.removeEventListener('touchend', delegatedTouchEndListener)
+    }
+  })
```

Comment:
As there are no dependencies to this effect, it will be run on each update of the component, which means the event listener will be removed and added on each update.

Can't `onTouchEnd` prop be used on the div element?

Reply:
I'd like to allow the user to drag his finger outside the panel and still have the gesture be registered properly, so sadly I have to listen to the whole page...
I could revert the component back to a class component to avoid this, unless there is a way to do it with hooks?

Reply:
Ok, I found a solution: having an empty array as a second argument for `useEffect` limits the call to this hook to the mount and unmount of the component, solving the issue :)

Source: https://stackoverflow.com/questions/53464595/how-to-use-componentwillmount-in-react-hooks

Reply:
Oh ok then maybe you can:

1. Use a Ref instead of a state as `fingerYPosition` value is only used in a futur `touchend` event.
  ```tsx
  const fingerYPosition = useRef<number | null>(null)
  // Reset it with:
  fingerYPosition.current = null
  ```

2. Move the disabled check at the top of the effect function, so you only add the event listener when you need it.
  ```tsx
  useEffect(() => {
    if (!disabled) {
      window.addEventListener()
      return () => {
        window.removeEventListener()
      }
    }
  }, [disabled]) // <= Still need to be a dependency
  ```

3. Put `onSlideUp` and `onSlideDown` in Refs so they can be removed from the dependencies while being defined inline.
  ```tsx
  const onSlideUpRef = useRef(onSlideUp)
  useEffect(() => (onSlideUpRef.current = onSlideUp)) // <= Need to be in an effect to make sure we only update the Ref when the component has been updated.

  onSlideUpRef.current() // <= Calls the "latest" reference of `onSlideUp`.

  <Grip
    onSlideUp={() => { /* ... */ }} // <= Won't remove/add the global event listener on each render.
  />
  ```